--- printer_ui-orig/printerui/qml/printer/PrintOptionsPage.qml
+++ printer_ui/printerui/qml/printer/PrintOptionsPage.qml
@@ -5,12 +5,18 @@
 
 import "qrc:/uibase/qml/widgets"
 import ".."
+import "../X1Plus.js" as X1Plus
 
 Item {
 
     property var printConfig: PrintManager.printConfig
     property var doorOpenLevelModel: [qsTr("notification"), qsTr("pause printing")]
-
+    Component.onCompleted: {
+        if (printConfig == {}){
+           // X1Plus.Bindings.updatePrintConfig();
+        }
+    }
+    
     ScrollView {
         id: scroll
         anchors.fill: parent
