--- printer_ui-orig/printerui/qml/printer/PrintListener.qml
+++ printer_ui/printerui/qml/printer/PrintListener.qml
@@ -1,15 +1,23 @@
 import QtQuick 2.12
 import Printer 1.0
+import "../X1Plus.js" as X1Plus
 
 Item {
 
     property bool printing: PrintManager.currentTask.stage >= PrintTask.WORKING
     property var printState: PrintManager.currentTask.state
-
+    property var totalLayer: X1Plus.Bindings.totalLayerNum()
+    property var layerNum: PrintManager.currentTask.layerNum
+   
     onPrintingChanged: {
         DeviceManager.power.externalWakeup()
     }
 
+    onLayerNumChanged: {
+        X1Plus.Bindings.printLayer.setCurrent();
+        X1Plus.Bindings.printLayer.setTotal();        
+    }
+    
     Shortcut {
         sequences: ["Stop", "Ctrl+B"]
         onActivated: {
